server/api/public.py

```diff
@@ -42,15 +47,106 @@
 # Specifying election results in a query param could cause us to hit URL size limits
 @api.route("/public/sample-sizes", methods=["POST"])
 def public_compute_sample_sizes():
-    compute_sample_sizes_input = request.get_json()
-    validate(compute_sample_sizes_input, COMPUTE_SAMPLE_SIZES_INPUT_SCHEMA)
-    sample_sizes = {
-        "ballotComparison": {},
-        "ballotPolling": {},
-        "batchComparison": {},
+    contest_for_sampler = parse_compute_sample_sizes_input(request.get_json())
+
+    audit_types = ["ballotComparison", "ballotPolling", "batchComparison"]
+    sample_sizes = {audit_type: {} for audit_type in audit_types}
+    sample_size_functions = {
+        "ballotComparison": compute_ballot_comparison_sample_size,
+        "ballotPolling": compute_ballot_polling_sample_size,
+        "batchComparison": compute_batch_comparison_sample_size,
     }
-    for risk_limit_percentage in list(range(0, 21)):
-        sample_sizes["ballotComparison"][str(risk_limit_percentage)] = 0
-        sample_sizes["ballotPolling"][str(risk_limit_percentage)] = 0
-        sample_sizes["batchComparison"][str(risk_limit_percentage)] = 0
+    for audit_type in audit_types:
+        for risk_limit_percentage in list(range(0, 21)):
+            sample_size_function = sample_size_functions[audit_type]
+            sample_sizes[audit_type][str(risk_limit_percentage)] = sample_size_function(
+                contest_for_sampler, risk_limit_percentage
+            )
+
     return jsonify(sample_sizes)
+
+
+def parse_compute_sample_sizes_input(
+    compute_sample_sizes_input: Any,
+) -> sampler_contest.Contest:
+    validate(compute_sample_sizes_input, COMPUTE_SAMPLE_SIZES_INPUT_SCHEMA)
+
+    election_results = compute_sample_sizes_input["electionResults"]
+    candidates = election_results["candidates"]
+    num_winners = election_results["numWinners"]
+    total_ballots_cast = election_results["totalBallotsCast"]
+
+    if all(candidate["votes"] == 0 for candidate in candidates):
+        raise Conflict("At least 1 candidate must have greater than 0 votes")
+
+    if num_winners >= len(candidates):
+        raise Conflict("Number of winners must be less than number of candidates")
+
+    candidate_votes_sum = sum(candidate["votes"] for candidate in candidates)
+    if total_ballots_cast < candidate_votes_sum:
+        raise Conflict("Total ballots cast cannot be less than sum of candidate votes")
+
+    contest = Contest(
+        choices=[
+            ContestChoice(
+                id=f"candidate-{i}",
```

Comment:
Should we require unique candidate names and use `candidate["name"]` here?

Reply:
I thought about this, but didn't see a huge benefit. I could (maybe) see a world where someone accidentally repeats candidate names in the original form, entering votes in like batches, e.g.

```
Helga Hippo | 10
Helga Hippo | 22
...
```

So this would prevent that. It's an easy enough change, so I'll make it.

Reply:
https://github.com/votingworks/arlo/pull/1600/commits/dc8030f03a5e188a156af600c9abc47534f8ee2f